File: zenspider/2018/01.rb
#!/usr/bin/env ruby -w

nums = File.readlines("01.txt").map(&:to_i)

p nums.sum

seen = Hash.new 0

p nums.cycle.reduce(0) { |freq, n|
  m = freq + n
  break m if (seen[m] += 1) > 1
  m
}

File: zenspider/2018/01.swift
#!/usr/bin/env swift -I .

import Foundation
import helpers

func parse(_ input: String) -> [Int] {
    return input.lines().mapInts()
}

func problem1(input: String) -> String {
    return parse(input).reduce(0, +).description
}

func problem2(input: String) -> String {
    var freq = 0
    var seen: Set<Int> = []

    for n in parse(input).cycle() { // using for as `break` doesn't work in `reduce`
        freq += n
        if seen.contains(freq) {
            break
        }
        seen.insert(freq)
    }

    return freq.description
}

func main() {
    interact([problem1, problem2])
}

main()
